### PR TITLE
API: make `hide_columns` and `hide_index` have a consistent signature and function in `Styler`

### PR DESCRIPTION
**Pull request by attack68**
This closes #41158 (which is an alternative PR for similar functionality)

Currently `hide_index()` and `hide_columns(subset)` have similar signatures but **different** operations. One hides the index whilst showing the data-rows, and the other hides data-columns whilst showing the column-headers row.

### This PR

 - adds the `subset` keyword to give: `hide_index(subset=None)`.
 - sets the default to `hide_columns(subset=None)`.

When `subset` is None the function now operates to hide the entire index, or entire column headers row respectively.
When `subset` is not None it operates to selectively hide the given rows or columns respectively, keeping the index or column headers row.

<s>We also add the `show` keyword to allow the method to operate inversely.</s>

@jreback I think you will prefer this over the `hide_values` and `hide_headers` methods suggested in #41158 since this re-uses the existing methods with minimal changes but tries to align their consistency. It is also fully backwards compatible.

### CODE REVIEW

File: pandas/io/formats/style_render.py [hunk 1]
```diff
@@ -209,37 +211,38 @@ def _translate_header(
...
-                    "th",
-                    f"{blank_class if name is None else index_name_class} level{r}",
-                    name if name is not None else blank_value,
-                    not self.hidden_index,
-                )
-            ]
-
-            if clabels:
-                column_headers = [
+        if not self.hidden_colheads:
```
Comment (attack68) [inline pandas/io/formats/style_render.py:214]: all that was changed here was to add `if not self.hidden_colheads:`, the other additions and removals in comparison window are misleading.

**Comment by jreback** [comment]
ok i undestand what you are doing, but this is still a very confusing api. we want to support

- hide index values themselves
- hide column values themselves
- hide particular columns

can you show me what you would write for these?

**Comment by attack68** [comment]
> ok i undestand what you are doing, but this is still a very confusing api. we want to support
> 
> * hide index values themselves
> * hide column values themselves
> * hide particular columns
> 
> can you show me what you would write for these?

sure..

Screen Shot 2021-05-03 at 18 53 22

The mechanics seem to be in place to do this quite easily, it just needs steering on the best API implementation..

**Comment by jreback** [comment]
ok your example looks fine. I am confused by the `show` keyword then.

**Comment by attack68** [comment]
> ok your example looks fine. I am confused by the `show` keyword then.

`df.style_hide_columns("b", show=True)` is equivalent to (the non-existant) `df.style.show_columns("b")`

Screenshot 2021-05-03 at 22 36 05

**Comment by attack68** [comment]
> ok your example looks fine. I am confused by the `show` keyword then.

I removed the `show` option. Can always add it back later or do it in a different way

File: pandas/io/formats/style.py [hunk 2]
```diff
@@ -1035,37 +1036,148 @@ def set_na_rep(self, na_rep: str) -> StylerRenderer:
...
+            subset = non_reducing_slice(subset)
+            hide = self.data.loc[subset]
+            hrows = self.index.get_indexer_for(hide.index)
+            # error: Incompatible types in assignment (expression has type
+            # "ndarray", variable has type "Sequence[int]")
+            self.hidden_rows = hrows  # type: ignore[assignment]
         return self
 
-    def hide_columns(self, subset) -> Styler:
+    def hide_columns(self, subset=None) -> Styler:
```
Comment (jreback) [inline pandas/io/formats/style.py:1109]: type subset & add versionchanged

File: pandas/io/formats/style.py [hunk 3]
```diff
@@ -1035,37 +1036,148 @@ def set_na_rep(self, na_rep: str) -> StylerRenderer:
...
+        This method has dual functionality:
+
+          - if ``subset`` is ``None`` then the entire column headers row will be hidden
+            whilst the data-values remain visible.
+          - if a ``subset`` is given then those specific columns, including the
+            data-values will be hidden, whilst the column headers row remains visible.
 
         Parameters
         ----------
         subset : IndexSlice
```
[the remark this answers is not in the document]
Comment (jreback) [reply pandas/io/formats/style.py:1122]: now I wouldn't object to *also* allowing an IndexSlice here (in addition to a list-like of columns labels)

File: pandas/io/formats/style_render.py [hunk 4]
```diff
@@ -98,6 +98,8 @@ def __init__(
 
         # add rendering variables
         self.hidden_index: bool = False
+        self.hidden_rows: Sequence[int] = []
```
Comment (jreback) [inline pandas/io/formats/style_render.py:101]: can you change the impl to use the same nomenclature, e.g. hidden_index, hidden_columns (or similar)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/formats/style_render.py (hunk 4 above)
Comment (attack68) [reply pandas/io/formats/style_render.py:101]: Can you give your input to the suggestion below:
```
self.hide_index_: bool
self.hide_columns_: bool
self.hidden_indexes: Sequence[int]
self.hidden_columns: Sequence[int]
```

File: pandas/io/formats/style.py [hunk 5]
```diff
@@ -1035,37 +1036,148 @@ def set_na_rep(self, na_rep: str) -> StylerRenderer:
...
+        This method has dual functionality:
+
+          - if ``subset`` is ``None`` then the entire index will be hidden whilst
+            displaying all data-rows.
+          - if a ``subset`` is given then those specific rows will be hidden whilst the
+            index itself remains visible.
+
+        Parameters
+        ----------
+        subset : IndexSlice
```
[the remark this answers is not in the document]
Comment (attack68) [reply pandas/io/formats/style.py]: changed. pls review

File: pandas/io/formats/style.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (attack68) [reply pandas/io/formats/style.py]: changed. pls review

File: pandas/io/formats/style.py [hunk 6]
```diff
@@ -1035,37 +1036,148 @@ def set_na_rep(self, na_rep: str) -> StylerRenderer:
...
+
+        See Also
+        --------
+        Styler.hide_index: Hide the entire index, or specific keys in the index.
+
+        Examples
+        --------
+        Hide column headers and retain the data values:
+
+        >>> midx = pd.MultiIndex.from_product([["x", "y"], ["a", "b", "c"]])
```
[the remark this answers is not in the document]
Comment (attack68) [reply pandas/io/formats/style.py:1179]: added

File: pandas/io/formats/style_render.py (hunk 4 above)
Comment (attack68) [reply pandas/io/formats/style_render.py]: updated (used rows instead of indexes)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
this is good, thanks @attack68 as a followup can you make sure that these methods are in the api ref & also we have sufficient docs in the notebook (the doc-strings are great obviously), but generally give an update in the notebook (its not of course necessary to add everything, but major things like this are good).

**Comment by jreback** [comment]
@meeseeksdev backport 1.3.x

**Comment by Leonardofreua** [comment]
Hi @attack68. I'm working on issue #42674 and I found some doctests with expected results different from the real results (at least in my tests). See the examples below:

* **hide_columns() - styles.py**
```
>>> midx = pd.MultiIndex.from_product([["x", "y"], ["a", "b", "c"]])
>>> df = pd.DataFrame(np.random.randn(6,6), index=midx, columns=midx)
>>> df.style.format("{:.1f}").hide_columns()
x   d    0.1    0.0    0.4    1.3    0.6   -1.4
    e    0.7    1.0    1.3    1.5   -0.0   -0.2
    f    1.4   -0.8    1.6   -0.2   -0.4   -0.3
y   d    0.4    1.0   -0.2   -0.8   -1.2    1.1
    e   -0.6    1.2    1.8    1.9    0.3    0.3
    f    0.8    0.5   -0.3    1.2    2.2   -0.8
```
As you can see below, the values `d`, `e`, and `f` don't appear in the result, because they were actually defined as `a`, `b`, and `c`.  We can also see that `x` and `y` are aligned with `b` (second and fifth lines) and not with `d` (first and third lines):
image

Is there any reason these results are divergent?

Obs.: There are some other doctests in the same method after the one mentioned above, which present the same problem.

**Pull request merged**